ofmipd.8
.TH ofmipd 8
.SH NAME
ofmipd \- accept outgoing mail through OFMIP
.SH SYNOPSIS
.B ofmipd
.I name.cdb
[
.I hostname
.I checkpasswd
.I truepgm
]
.SH DESCRIPTION
.B ofmipd
accepts mail messages through the
Old-Fashioned Mail Injection Protocol (OFMIP),
rewrites them according to local rules,
and invokes
.B qmail-queue
to deposit them into the outgoing queue.
.B ofmipd
must be supplied several environment variables;
see
.BR tcp-environ (5).

See
.B http://pobox.com/~djb/proto/ofmip.html
for further information on OFMIP.
.SH "USING OFMIP"
You can use
.B tcpserver
to set up
.B ofmipd
on a free TCP port.
Some sites use port 26.
Some sites use port 25 on an IP address that does not receive incoming mail.

The
.I hostname
argument is the name of the mail host, intended to create CRAM password
challenges but currently ignored.
The
.I checkpasswd
argument
is the name of a password checker that uses the same calling conventions as
the qmail POP3 checkpasswd.
The
.I truepgm
argument is the name a program that succeeds, such as
.B /bin/true\c
, for POP3 checkpasswd compatibility.

If the environment variable RELAYCLIENT is set,
or the client logs in with AUTH,
.B ofmipd
will relay messages to any destination.
If not, it will reject any MAIL FROM, RCPT TO, or DATA command with a 503
error.
With
.B tcpserver
you can set RELAYCLIENT for 
preauthorized IP addresses such as 127.0.0.1.
The contents of RELAYCLIENT are ignored; in particular, they are not
appended to destnation mail addresses.

Most MUAs that claim to be ``SMTP clients''
are actually OFMIP clients.
You should point them at
.B ofmipd
rather than
.BR qmail-smtpd .
Some MUAs will use port 26 of 
.I server
if you tell them that the ``SMTP server'' is
.IR server\fB:26 .
.SH "MESSAGE REWRITING"
.B ofmipd
rewrites each message in essentially the same way that
.B new-inject
does.
It transforms each envelope address
as discussed in
.BR rewriting (5);
it cleans up address lists in
.BR To ,
.BR Cc ,
.BR Notice-Requested-Upon-Delivery-To ,
.BR From ,
.BR Sender ,
.BR Reply-To ,
.BR Mail-Reply-To ,
and
.BR Mail-Followup-To ;
it adds
.B Date
and
.BR Message-ID ;
it discards
.BR Bcc ,
.BR Apparently-To ,
.BR Content-Length ,
and
.BR Return-Path ;
and it moves various fields to the top of the message.

.B ofmipd
also transforms envelope sender addresses and
.B From
lines according to
.I name.cdb
if
.I name.cdb
is supplied.
See
.BR ofmipname (8)
for further details.
The
.I name.cdb
argument must be supplied but may be a null string of no name
transformation file is to be used.

.B ofmipd
accepts LF and CR LF as line terminators inside messages.

.B ofmipd
does not do hop counting
or local-IP-address recognition.
.SH "SEE ALSO"
new-inject(1),
tcpserver(1),
addresses(5),
tcp-environ(5),
rewriting(5),
qmail-control(5),
qmail-queue(8),
ofmipname(8)


ofmipd.c
#include "commands.h"
#include "sig.h"
#include "auto_qmail.h"
#include "qmail.h"
#include "readwrite.h"
#include "timeoutread.h"
#include "timeoutwrite.h"
#include "stralloc.h"
#include "substdio.h"
#include "config.h"
#include "env.h"
#include "exit.h"
#include "error.h"
#include "str.h"
#include "mess822.h"
#include "tai.h"
#include "caltime.h"
#include "cdb.h"
#include "base64.h"
#include "wait.h"
#include "fd.h"
#include "byte.h"
#include "case.h"
  
#undef AUTHCRAM			/* don't define, not fully implemented */

int timeout = 1200;

char *relayclient;

int safewrite(fd,buf,len) int fd; char *buf; int len;
{
  int r;
  r = timeoutwrite(timeout,fd,buf,len);
  if (r <= 0) _exit(1);
  return r;
}

char ssoutbuf[512];
substdio ssout = SUBSTDIO_FDBUF(safewrite,1,ssoutbuf,sizeof ssoutbuf);

void flush() { substdio_flush(&ssout); }
void out(s) char *s; { substdio_puts(&ssout,s); }

void die_read() { _exit(1); }
void nomem() { out("451 out of memory (#4.3.0)\r\n"); flush(); _exit(1); }
void die_config() { out("451 unable to read configuration (#4.3.0)\r\n"); flush(); _exit(1); }
void smtp_quit() { out("221 ofmipd.local\r\n"); flush(); _exit(0); }
void smtp_help() { out("214 qmail home page: http://pobox.com/~djb/qmail.html\r\n"); }
void smtp_noop() { out("250 ok\r\n"); }
void smtp_vrfy() { out("252 send some mail, i'll try my best\r\n"); }
void smtp_unimpl() { out("502 unimplemented (#5.5.1)\r\n"); }
void err_syntax() { out("555 syntax error (#5.5.4)\r\n"); }
void err_wantmail() { out("503 MAIL first (#5.5.1)\r\n"); }
void err_wantrcpt() { out("503 RCPT first (#5.5.1)\r\n"); }
void err_qqt() { out("451 qqt failure (#4.3.0)\r\n"); }
void err_cdb() { out("451 unable to read cdb (#4.3.0)\r\n"); }
int err_child() { out("454 oops, problem with child and I can't auth (#4.3.0)\r\n"); return -1; }
int err_fork() { out("454 oops, child won't start and I can't auth (#4.3.0)\r\n"); return -1; }
int err_pipe() { out("454 oops, unable to open pipe and I can't auth (#4.3.0)\r\n"); return -1; }
int err_write() { out("454 oops, unable to write pipe and I can't auth (#4.3.0)\r\n"); return -1; }
void err_authd() { out("503 you're already authenticated (#5.5.0)\r\n"); }
void err_authmail() { out("503 no auth during mail transaction (#5.5.0)\r\n"); }
int err_noauth() { out("504 auth type unimplemented (#5.5.1)\r\n"); return -1; }
int err_authabrt() { out("501 auth exchange cancelled (#5.0.0)\r\n"); return -1; }
int err_input() { out("501 malformed auth input (#5.5.4)\r\n"); return -1; }
int err_notauth() { out("503 authorize or check your mail before sending (#5.5.1)\r\n"); return -1; }

config_str rewrite = CONFIG_STR;
stralloc idappend = {0};

stralloc addr = {0}; /* will be 0-terminated, if addrparse returns 1 */
stralloc rwaddr = {0};

int addrparse(arg)
char *arg;
{
  int i;
  char ch;
  char terminator;
  int flagesc;
  int flagquoted;
 
  terminator = '>';
  i = str_chr(arg,'<');
  if (arg[i])
    arg += i + 1;
  else { /* partner should go read rfc 821 */
    terminator = ' ';
    arg += str_chr(arg,':');
    if (*arg == ':') ++arg;
    while (*arg == ' ') ++arg;
  }

  if (*arg == '@') while (*arg) if (*arg++ == ':') break;

  if (!stralloc_copys(&addr,"")) nomem();
  flagesc = 0;
  flagquoted = 0;
  for (i = 0;ch = arg[i];++i) { /* copy arg to addr, stripping quotes */
    if (flagesc) {
      if (!stralloc_append(&addr,&ch)) nomem();
      flagesc = 0;
    }
    else {
      if (!flagquoted && (ch == terminator)) break;
      switch(ch) {
        case '\\': flagesc = 1; break;
        case '"': flagquoted = !flagquoted; break;
        default: if (!stralloc_append(&addr,&ch)) nomem();
      }
    }
  }

  if (!rewritehost_addr(&rwaddr,addr.s,addr.len,config_data(&rewrite))) nomem();

  return rwaddr.len < 900;
}

char *fncdb;
int fdcdb;
stralloc cdbresult = {0};

int seenmail = 0;
char *name; /* defined if seenmail; points into cdbresult */

stralloc mailfrom = {0};
stralloc rcptto = {0};

void smtp_helo(arg) char *arg;
{
  seenmail = 0;
  out("250 ofmipd.local\r\n");
}
void smtp_ehlo(arg) char *arg;
{
  seenmail = 0;
  out("250-ofmipd.local");
#ifdef AUTHCRAM
  out("\r\n250-AUTH LOGIN CRAM-MD5 PLAIN");
  out("\r\n250-AUTH=LOGIN CRAM-MD5 PLAIN");
#else
  out("\r\n250-AUTH LOGIN PLAIN");
  out("\r\n250-AUTH=LOGIN PLAIN");
#endif
  out("\r\n250-PIPELINING\r\n250 8BITMIME\r\n");
}
void smtp_rset()
{
  seenmail = 0;
  out("250 flushed\r\n");
}
void smtp_mail(arg) char *arg;
{
  if (!relayclient) { err_notauth(); return; }
  if (!addrparse(arg)) { err_syntax(); return; }

  name = 0;
  if (fncdb) {
    uint32 dlen;
    int r;

    r = cdb_seek(fdcdb,rwaddr.s,rwaddr.len,&dlen);
    if (r == -1) { err_cdb(); return; }
    if (r) {
      if (!stralloc_ready(&cdbresult,(unsigned int) dlen)) nomem();
      cdbresult.len = dlen;
      name = cdbresult.s;
      if (cdb_bread(fdcdb,name,cdbresult.len) == -1) { err_cdb(); return; }
      r = byte_chr(name,cdbresult.len,'\0');
      if (r == cdbresult.len) { err_cdb(); return; }
      if (!stralloc_copyb(&rwaddr,cdbresult.s + r + 1,cdbresult.len - r - 1)) nomem();
    }
  }

  if (!stralloc_copy(&mailfrom,&rwaddr)) nomem();
  if (!stralloc_0(&mailfrom)) nomem();
  if (!stralloc_copys(&rcptto,"")) nomem();
  seenmail = 1;
  out("250 ok\r\n");
}
void smtp_rcpt(arg) char *arg; {
  if (!relayclient) { err_notauth(); return; }
  if (!seenmail) { err_wantmail(); return; }
  if (!addrparse(arg)) { err_syntax(); return; }
  if (!stralloc_0(&rwaddr)) nomem();
  if (!stralloc_cats(&rcptto,"T")) nomem();
  if (!stralloc_cats(&rcptto,rwaddr.s)) nomem();
  if (!stralloc_0(&rcptto)) nomem();
  out("250 ok\r\n");
}

struct qmail qqt;
void put(buf,len) char *buf; int len; { qmail_put(&qqt,buf,len); }
void puts(buf) char *buf; { qmail_puts(&qqt,buf); }

stralloc tmp = {0};
stralloc tmp2 = {0};

void rewritelist(list)
stralloc *list;
{
  if (!rewritehost_list(&tmp,list->s,list->len,config_data(&rewrite))) nomem();
  if (!stralloc_copy(list,&tmp)) nomem();
}

void putlist(name,list)
char *name;
stralloc *list;
{
  if (!list->len) return;
  if (!mess822_quotelist(&tmp,list)) nomem();
  if (!mess822_fold(&tmp2,&tmp,name,78)) nomem();
  put(tmp2.s,tmp2.len);
}

mess822_time datastart;
stralloc datastamp = {0};

mess822_time date;
stralloc to = {0};
stralloc cc = {0};
stralloc nrudt = {0};
stralloc from = {0};
stralloc headersender = {0};
stralloc replyto = {0};
stralloc mailreplyto = {0};
stralloc followupto = {0};

stralloc msgid = {0};
stralloc top = {0};
stralloc bottom = {0};

mess822_header h = MESS822_HEADER;
mess822_action a[] = {
  { "date", 0, 0, 0, 0, &date }
, { "to", 0, 0, 0, &to, 0 }
, { "cc", 0, 0, 0, &cc, 0 }
, { "notice-requested-upon-delivery-to", 0, 0, 0, &nrudt, 0 }
, { "from", 0, 0, 0, &from, 0 }
, { "sender", 0, 0, 0, &headersender, 0 }
, { "reply-to", 0, 0, 0, &replyto, 0 }
, { "mail-reply-to", 0, 0, 0, &mailreplyto, 0 }
, { "mail-followup-to", 0, 0, 0, &followupto, 0 }
, { "message-id", 0, &msgid, 0, 0, 0 }
, { "received", 0, &top, 0, 0, 0 }
, { "delivered-to", 0, &top, 0, 0, 0 }
, { "errors-to", 0, &top, 0, 0, 0 }
, { "return-receipt-to", 0, &top, 0, 0, 0 }
, { "resent-sender", 0, &top, 0, 0, 0 }
, { "resent-from", 0, &top, 0, 0, 0 }
, { "resent-reply-to", 0, &top, 0, 0, 0 }
, { "resent-to", 0, &top, 0, 0, 0 }
, { "resent-cc", 0, &top, 0, 0, 0 }
, { "resent-bcc", 0, &top, 0, 0, 0 }
, { "resent-date", 0, &top, 0, 0, 0 }
, { "resent-message-id", 0, &top, 0, 0, 0 }
, { "bcc", 0, 0, 0, 0, 0 }
, { "return-path", 0, 0, 0, 0, 0 }
, { "apparently-to", 0, 0, 0, 0, 0 }
, { "content-length", 0, 0, 0, 0, 0 }
, { 0, 0, &bottom, 0, 0, 0 }
} ;

void finishheader()
{
  if (!mess822_end(&h)) nomem();

  if (name) from.len = 0;

  rewritelist(&to);
  rewritelist(&cc);
  rewritelist(&nrudt);
  rewritelist(&from);
  rewritelist(&headersender);
  rewritelist(&replyto);
  rewritelist(&mailreplyto);
  rewritelist(&followupto);

  put(top.s,top.len);

  if (!date.known) date = datastart;
  if (!mess822_date(&tmp,&date)) nomem();
  puts("Date: ");
  put(tmp.s,tmp.len);
  puts("\n");

  if (!msgid.len) {
    static int idcounter = 0;

    if (!stralloc_copys(&msgid,"Message-ID: <")) nomem();
    if (!stralloc_catlong(&msgid,date.ct.date.year)) nomem();
    if (!stralloc_catint0(&msgid,date.ct.date.month,2)) nomem();
    if (!stralloc_catint0(&msgid,date.ct.date.day,2)) nomem();
    if (!stralloc_catint0(&msgid,date.ct.hour,2)) nomem();
    if (!stralloc_catint0(&msgid,date.ct.minute,2)) nomem();
    if (!stralloc_catint0(&msgid,date.ct.second,2)) nomem();
    if (!stralloc_cats(&msgid,".")) nomem();
    if (!stralloc_catint(&msgid,++idcounter)) nomem();
    if (!stralloc_cat(&msgid,&idappend)) nomem();
    if (!stralloc_cats(&msgid,">\n")) nomem();
  }
  put(msgid.s,msgid.len);

  putlist("From: ",&from);
  if (!from.len) {
    puts("From: ");
    if (!mess822_quote(&tmp,mailfrom.s,name)) nomem();
    put(tmp.s,tmp.len);
    puts("\n");
  }

  putlist("Sender: ",&headersender);
  putlist("Reply-To: ",&replyto);
  putlist("Mail-Reply-To: ",&mailreplyto);
  putlist("Mail-Followup-To: ",&followupto);
  if (!to.len && !cc.len)
    puts("Cc: recipient list not shown: ;\n");
  putlist("To: ",&to);
  putlist("Cc: ",&cc);
  putlist("Notice-Requested-Upon-Delivery-To: ",&nrudt);

  put(bottom.s,bottom.len);
}

int saferead(fd,buf,len) int fd; char *buf; int len;
{
  int r;
  flush();
  r = timeoutread(timeout,fd,buf,len);
  if (r <= 0) die_read();
  return r;
}

char ssinbuf[1024];
substdio ssin = SUBSTDIO_FDBUF(saferead,0,ssinbuf,sizeof ssinbuf);

stralloc line = {0};
int match;

void blast()
{
  int flagheader = 1;
  int i;

  if (!mess822_begin(&h,a)) nomem();

  for (;;) {
    if (getln(&ssin,&line,&match,'\n') == -1) die_read();
    if (!match) die_read();

    --line.len;
    if (line.len && (line.s[line.len - 1] == '\r')) --line.len;
    if (line.len && (line.s[0] == '.')) {
      --line.len;
      if (!line.len) break;
      for (i = 0;i < line.len;++i) line.s[i] = line.s[i + 1];
    }
    line.s[line.len++] = '\n';

    if (flagheader)
      if (!mess822_ok(&line)) {
        finishheader();
	flagheader = 0;
	if (line.len > 1) put("\n",1);
      }
    if (!flagheader)
      put(line.s,line.len);
    else
      if (!mess822_line(&h,&line)) nomem();
  }

  if (flagheader)
    finishheader();
}

void safecats(out,in)
stralloc *out;
char *in;
{
  char ch;
  while (ch = *in++) {
    if (ch < 33) ch = '?';
    if (ch > 126) ch = '?';
    if (ch == '(') ch = '?';
    if (ch == ')') ch = '?';
    if (ch == '@') ch = '?';
    if (ch == '\\') ch = '?';
    if (!stralloc_append(out,&ch)) nomem();
  }
}

stralloc received = {0};
char *remoteinfo;

void received_init()
{
  char *x;

  if (!stralloc_copys(&received,"Received: (ofmipd ")) nomem();
  x = remoteinfo;
  if (x) {
    safecats(&received,x);
    if (!stralloc_append(&received,"@")) nomem();
  }
  x = env_get("TCPREMOTEIP");
  if (!x) x = "unknown";
  safecats(&received,x);
  if (!stralloc_cats(&received,"); ")) nomem();
}

void smtp_data() {
  struct tai now;
  char *qqx;

  tai_now(&now);
  caltime_utc(&datastart.ct,&now,(int *) 0,(int *) 0);
  datastart.known = 1;
  if (!mess822_date(&datastamp,&datastart)) nomem();
 
  if (!relayclient) { err_notauth(); return; }
  if (!seenmail) { err_wantmail(); return; }
  if (!rcptto.len) { err_wantrcpt(); return; }
  seenmail = 0;
  if (qmail_open(&qqt) == -1) { err_qqt(); return; }
  out("354 go ahead\r\n");
 
  received_init();
  qmail_put(&qqt,received.s,received.len);
  qmail_put(&qqt,datastamp.s,datastamp.len);
  qmail_puts(&qqt,"\n");
  blast();
  qmail_from(&qqt,mailfrom.s);
  qmail_put(&qqt,rcptto.s,rcptto.len);
 
  qqx = qmail_close(&qqt);
  if (!*qqx) { out("250 ok\r\n"); return; }
  if (*qqx == 'D') out("554 "); else out("451 ");
  out(qqx + 1);
  out("\r\n");
}

#ifdef AUTHCRAM
char unique[FMT_ULONG + FMT_ULONG + 3];
#endif
static stralloc authin = {0};
static stralloc user = {0};
static stralloc pass = {0};
static stralloc resp = {0};
static stralloc slop = {0};
char *hostname;
char **childargs;
substdio ssup;
char upbuf[128];
int authd = 0;

int authgetl(void) {
  int i;

  if (!stralloc_copys(&authin, "")) nomem();

  for (;;) {
    if (!stralloc_readyplus(&authin,1)) nomem(); /* XXX */
    i = substdio_get(&ssin,authin.s + authin.len,1);
    if (i != 1) die_read();
    if (authin.s[authin.len] == '\n') break;
    ++authin.len;
  }

  if (authin.len > 0) if (authin.s[authin.len - 1] == '\r') --authin.len;
  authin.s[authin.len] = 0;

  if (*authin.s == '*' && *(authin.s + 1) == 0) { return err_authabrt(); }
  if (authin.len == 0) { return err_input(); }
  return authin.len;
}

int authenticate(void)
{
  int child;
  int wstat;
  int pi[2];

  if (!stralloc_0(&user)) nomem();
  if (!stralloc_0(&pass)) nomem();
  if (!stralloc_0(&resp)) nomem();

  if (fd_copy(2,1) == -1) return err_pipe();
  close(3);
  if (pipe(pi) == -1) return err_pipe();
  if (pi[0] != 3) return err_pipe();
  switch(child = fork()) {
    case -1:
      return err_fork();
    case 0:
      close(pi[1]);
      sig_pipedefault();
      execvp(*childargs, childargs);
      _exit(1);
  }
  close(pi[0]);

  substdio_fdbuf(&ssup,write,pi[1],upbuf,sizeof upbuf);
  if (substdio_put(&ssup,user.s,user.len) == -1) return err_write();
  if (substdio_put(&ssup,pass.s,pass.len) == -1) return err_write();
  if (substdio_put(&ssup,resp.s,resp.len) == -1) return err_write();
  if (substdio_flush(&ssup) == -1) return err_write();

  close(pi[1]);
  byte_zero(pass.s,pass.len);
  byte_zero(upbuf,sizeof upbuf);
  if (wait_pid(&wstat,child) == -1) return err_child();
  if (wait_crashed(wstat)) return err_child();
  if (wait_exitcode(wstat)) { sleep(5); return 1; } /* no */
  return 0; /* yes */
}

int auth_login(arg) char *arg;
{
  int r;

  if (*arg) {
    if (r = b64decode(arg,str_len(arg),&user) == 1) return err_input();
  }
  else {
    out("334 VXNlcm5hbWU6\r\n"); flush(); /* Username: */
    if (authgetl() < 0) return -1;
    if (r = b64decode(authin.s,authin.len,&user) == 1) return err_input();
  }
  if (r == -1) nomem();

  out("334 UGFzc3dvcmQ6\r\n"); flush(); /* Password: */

  if (authgetl() < 0) return -1;
  if (r = b64decode(authin.s,authin.len,&pass) == 1) return err_input();
  if (r == -1) nomem();

  if (!user.len || !pass.len) return err_input();
  return authenticate();  
}

int auth_plain(arg) char *arg;
{
  int r, id = 0;

  if (*arg) {
    if (r = b64decode(arg,str_len(arg),&slop) == 1) return err_input();
  }
  else {
    out("334 \r\n"); flush();
    if (authgetl() < 0) return -1;
    if (r = b64decode(authin.s,authin.len,&slop) == 1) return err_input();
  }
  if (r == -1 || !stralloc_0(&slop)) nomem();
  while (slop.s[id]) id++; /* ignore authorize-id */

  if (slop.len > id + 1)
    if (!stralloc_copys(&user,slop.s + id + 1)) nomem();
  if (slop.len > id + user.len + 2)
    if (!stralloc_copys(&pass,slop.s + id + user.len + 2)) nomem();

  if (!user.len || !pass.len) return err_input();
  return authenticate();
}

#ifdef AUTHCRAM
int auth_cram()
{
  int i, r;
  char *s;

  s = unique;
  s += fmt_uint(s,getpid());
  *s++ = '.';
  s += fmt_ulong(s,(unsigned long) now());
  *s++ = '@';
  *s++ = 0;

  if (!stralloc_copys(&pass,"<")) nomem();
  if (!stralloc_cats(&pass,unique)) nomem();
  if (!stralloc_cats(&pass,hostname)) nomem();
  if (!stralloc_cats(&pass,">")) nomem();
  if (b64encode(&pass,&slop) < 0) nomem();
  if (!stralloc_0(&slop)) nomem();

  out("334 ");
  out(slop.s);
  out("\r\n");
  flush();

  if (authgetl() < 0) return -1;
  if (r = b64decode(authin.s,authin.len,&slop) == 1) return err_input();
  if (r == -1 || !stralloc_0(&slop)) nomem();

  i = str_chr(slop.s,' ');
  s = slop.s + i;
  while (*s == ' ') ++s;
  slop.s[i] = 0;
  if (!stralloc_copys(&user,slop.s)) nomem();
  if (!stralloc_copys(&resp,s)) nomem();

  if (!user.len || !resp.len) return err_input();
  return authenticate();
}
#endif

struct authcmd {
  char *text;
  int (*fun)();
} authcmds[] = {
  { "login", auth_login }
, { "plain", auth_plain }
#ifdef AUTHCRAM
, { "cram-md5", auth_cram }
#endif
, { 0, err_noauth }
};

void smtp_auth(arg)
char *arg;
{
  int i;
  char *cmd = arg;

  if (!hostname || !*childargs)
  {
    out("503 auth not available (#5.3.3)\r\n");
    return;
  }
  if (authd) { err_authd(); return; }
  if (seenmail) { err_authmail(); return; }

  if (!stralloc_copys(&user,"")) nomem();
  if (!stralloc_copys(&pass,"")) nomem();
  if (!stralloc_copys(&resp,"")) nomem();

  i = str_chr(cmd,' ');   
  arg = cmd + i;
  while (*arg == ' ') ++arg;
  cmd[i] = 0;

  for (i = 0;authcmds[i].text;++i)
    if (case_equals(authcmds[i].text,cmd)) break;

  switch (authcmds[i].fun(arg)) {
    case 0:
      authd = 1;
      relayclient = "";
      remoteinfo = user.s;
      if (!env_unset("TCPREMOTEINFO")) die_read();
      if (!env_put2("TCPREMOTEINFO",remoteinfo)) nomem();
      out("235 ok, go ahead (#2.0.0)\r\n");
      break;
    case 1:
      out("535 authorization failed (#5.7.0)\r\n");
  }
}

struct commands smtpcommands[] = {
  { "rcpt", smtp_rcpt, 0 }
, { "mail", smtp_mail, 0 }
, { "data", smtp_data, flush }
, { "auth", smtp_auth, flush }
, { "quit", smtp_quit, flush }
, { "helo", smtp_helo, flush }
, { "ehlo", smtp_ehlo, flush }
, { "rset", smtp_rset, 0 }
, { "help", smtp_help, flush }
, { "noop", smtp_noop, flush }
, { "vrfy", smtp_vrfy, flush }
, { 0, smtp_unimpl, flush }
} ;

void main(argc,argv)
int argc;
char **argv;
{
  sig_pipeignore();

  fncdb = argv[1];
  if (fncdb && *fncdb) {
    fdcdb = open_read(fncdb);
    if (fdcdb == -1) die_config();
  } else fncdb = 0;

  hostname = argv[2];
  childargs = argv + 3;
  remoteinfo = env_get("TCPREMOTEINFO");
  relayclient = env_get("RELAYCLIENT");

  if (leapsecs_init() == -1) die_config();
  if (chdir(auto_qmail) == -1) die_config();
  if (rwhconfig(&rewrite,&idappend) == -1) die_config();

  out("220 ofmipd.local ESMTP\r\n");
  commands(&ssin,&smtpcommands);
  nomem();
}

